### ISSUE TITLE
Confirmed user error on first sign attempt, but not on second and so on. 

### ISSUE THREAD

**Issue by razrinn** [furaiev/amazon-cognito-identity-dart-2#192]
Hi @furaiev, this issue is related to #165.

Confirmed user case:
1. Tried to sign in for the first time
2. Got thrown an error `CognitoUserConfirmationNecessaryException` with user's session inside the error object, full error log:
```
CognitoUserException: User Confirmation Necessary
```
3. Tried to sign in again
4. The error is now gone

Actual unconfirmed user:
1. Tried to sign in
2. Got thrown an error `CognitoClientException` instead, full error log:
```
CognitoClientException{statusCode: 400, code: UserNotConfirmedException, name: UserNotConfirmedException, message: User is not confirmed.}
```

**Comment by byshy** [comment]
Facing same issue

**Comment by CristopherVidalMachado** [comment]
Same issue here!

**Comment by furaiev** [comment]
Hello, guys, you are welcome to debug and fix this bug if you have time and passion.
Unfortunately, I'm overloaded and don't have time to do it.

**Comment by anthonybouton** [comment]
This appears to be normal.
I'm bumping into this "issue" also, however it seems this works as intended.

It actually has nothing to do with the User not being confirmed, the exception being thrown here is that the Users it's current device has not yet been confirmed ( bit unlucky naming on the exceptions ).

Whenever a new device is detected, depending on the settings on your user pool, you might have your users verify the logon on the new device.
After one attempt, some device keys get cached which makes it "work" on the second time, again, depending on your user pool settings.

See
See https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ConfirmDevice.html 
and
https://aws.amazon.com/premiumsupport/knowledge-center/cognito-user-pool-remembered-devices/


TLDR.
Either fix the exception and confirm the device as intended on an already verified device
OR
Disable Device tracking on your user pool settings ( Don't remember devices )
OR
Always remember devices

When User Opt-in, you will bump into this "issue", which is not an issue :)

![image](https://user-images.githubusercontent.com/17692228/226140680-458e5fd3-719f-4d80-8dd5-71f31512fb11.png)

**Comment by furaiev** [comment]
@anthonybouton thank you for your contribution, your changes are available as of 3.3.0

**Comment by felipeapfer** [comment]
Hello!
I`m running on this "issue". When User Opt-in is activated the ConfirmUser exception is thrown and I do believe it is due to the fact that  this [PR](https://github.com/furaiev/amazon-cognito-identity-dart-2/pull/227) only throws CognitoUserDeviceConfirmationNecessaryException for MFA. 

In this scenario I believe it would be correct to throw the CognitoUserDeviceConfirmationNecessaryException too since it is a valid option configuration of Cognito.

As a work around one can check if  there is a valid signInUserSession on the Exception object. If so it is an issue with the device.
